### ISSUE TITLE
windows 10 linux subsystem support

### ISSUE THREAD

**Issue by denfromufa** [appveyor/ci#1295]
this is also called WSL or Bash on Windows.

**Comment by denfromufa** [comment]
It is now possible to install linux subsystem on preview builds of Windows Server 2016:

https://msdn.microsoft.com/en-us/commandline/wsl/install-on-server

**Comment by IlyaFinkelshteyn** [comment]
Hi Denis, this OS is still in beta as I understand. Actually we are going to support Linux VMs for build in near future, so you could run Windows and Linux builds side-by-side. Would it be better alternative?

**Comment by sjackman** [comment]
I'm interested in testing our command line tool on WSL. I already have CI set up to test it on Linux.

**Comment by IlyaFinkelshteyn** [comment]
This will be available when this become part of generally available Windows Server. Does anyone knows if this is going to be only with "Core" SKU or with UI-based version too?

Another option (if you need this ASAP) is to you own build worker image with [private build cloud](https://www.appveyor.com/docs/build-environment/#private-build-cloud) feature, but this feature is available for Premium accounts only right now.

**Comment by riban-bw** [comment]
Appveyor now supports Server 2016. Is it time to review this request?

**Comment by BR0kEN-** [comment]
Current build is 14393 and it would be awesome if there be a possibility to update it to, at least, 16215.

**Comment by felixfbecker** [comment]
Is there a workaround? Is it possible to install WSL as part of the build?

**Comment by FeodorFitsner** [comment]
@felixfbecker what's business case for having WSL on build workers? Could you please elaborate?

**Comment by jhecking** [comment]
@FeodorFitsner, for me at least, the business case for WSL is the same as that for [supporting Linux docker containers on Windows](https://github.com/appveyor/ci/issues/1717#issuecomment-356869863): I am developing a client SDK that needs to run on Linux and Windows, but the server only runs on Linux. In order to run integration tests for the client on Windows, I need to be able to spin up a server instance, using either WSL or a Linux docker container instance. Solving either of these two issues would solve my use case.

**Comment by FeodorFitsner** [comment]
Ah, nice, this is what I wanted to hear 😄

**Comment by felixfbecker** [comment]
@FeodorFitsner In my case, I want to test that a CLI works correctly inside WSL. Actual Linux CI wouldn't solve this (otherwise I could just use Travis).

**Comment by chuckatkins** [comment]
> what's business case for having WSL on build workers? Could you please elaborate?

I would very much like to have this.  We explicitly target the engineering and scientific research communities.  Most of our users work remotely on large Linux clusters but when developing locally on their laptop, they're usually using Windows on a machine issued by their organization.  As such, some of them chose to just natively use our tools in a Windows development environment while others just use WSL in order to continue to use a familiar preferred environment.  It effectively removes much of the use case for Cygwin and/or MinGW environments.

So in short, having WSL available on the build workers is useful because it's a widely used environment by my customers.

**Comment by FeodorFitsner** [comment]
@NickCraver we offer "premium" build VMs with Docker and nested virtualization enabled, so you can spin off Linux containers inside Windows build: https://ci.appveyor.com/project/appveyor-tests/docker-ce

That offer is not free for OSS though, but will cost $50/job/month.

**Comment by felixfbecker** [comment]
@FeodorFitsner I don't think that would solve our case - we want to test applications _specifically WSL_ to make sure they run properly in WSL, to make sure users/customers can run the application in WSL (which is not a given just because it runs in Linux, because WSL has many caveats/weirdnesses).

**Comment by FeodorFitsner** [comment]
OK, the Windows 10 workers is the way to go.

**Comment by jonashackt** [comment]
Is there any progress on the WSL support?

**Comment by billziss-gh** [comment]
@FeodorFitsner any progress on this?

Being able to test against WSL would be very useful for my project [WinFsp](https://github.com/billziss-gh/winfsp), because WinFsp file systems can now be mounted in WSL. I would like to add tests to regularly test this functionality via AppVeyor CI.

It is my understanding that [Windows Server supports WSL](https://docs.microsoft.com/en-us/windows/wsl/install-on-server) (as mentioned earlier) and I could perhaps write scripts to enable/install WSL, but it would be even better if the functionality came preinstalled.

**Comment by IlyaFinkelshteyn** [comment]
We recently introduced `Windows Server 2019` image, which has Docker for Windows and WSL to implement any kind of hybrid scenarios. Please see [usage example](https://ci.appveyor.com/project/appveyor-tests/win2k19-azure/builds/24024378).

Note that this image is running on Azure so be prepared for 3-4 minutes `Starting` stage (time to fire an Azure VM).

This image is available per request for now. @billziss-gh it is enabled for you.

If anybody else want to try this image, pleas ping us here or send a email to team@appveyor.com

**Comment by billziss-gh** [comment]
@IlyaFinkelshteyn thank you, much appreciated!

**Comment by alexfmpe** [comment]
@IlyaFinkelshteyn I'd like to try out the WS2019 image, if possible.

I've been meaning to get a WSL CI going for [reflex-platform](https://github.com/reflex-frp/reflex-platform/), for ages, since we depend on [nix](https://github.com/NixOS/nix) which doesn't run natively on Window. Some people have reported success with our project under WSL, but because there's far less users, it's hard to be aware when WSL in particular breaks, or to claim any kind of support for it.

**Comment by IlyaFinkelshteyn** [comment]
@alexfmpe what is your AppVeyor account?

**Comment by alexfmpe** [comment]
`alexteves13469` I think. It was linked to my github account, but showing an old (2-3 years) old username, for which the redirect no longer works. Re-linked properly now.

**Comment by IlyaFinkelshteyn** [comment]
@alexfmpe OK, it has been enabled

**Comment by alexfmpe** [comment]
Yop, I see it now. Thanks a lot

**Comment by juju4** [comment]
@IlyaFinkelshteyn interested too to test! Thanks a lot!

**Comment by IlyaFinkelshteyn** [comment]
@juju4 sure, enabled :)

**Comment by xeijin** [comment]
@IlyaFinkelshteyn please could you enable for my AppVeyor account also? Same username as here. Many thanks!

**Comment by IlyaFinkelshteyn** [comment]
@xeijin done

**Comment by stephengroat** [comment]
any hope of getting the `Windows Server 2019` on the `stephengroat` account? want to test wsl for haskall and `shellcheck`

**Comment by IlyaFinkelshteyn** [comment]
@stephengroat sure, enabled. Let us know if you have any issues.

**Comment by patrick330602** [comment]
Is it still possible to get the `Windows Server 2019` for CI? I hope to test my WSL-specific project with `bats` on WSL. If it is still possible, my account is `patrick330602`.

**Comment by IlyaFinkelshteyn** [comment]
@patrick330602 sure, done.

**Comment by chuckatkins** [comment]
@IlyaFinkelshteyn can you please enable WSL for the ornladios/adios2 account / project that I admin?

**Comment by FeodorFitsner** [comment]
Hey @chuckatkins, we've just deployed an updated to `Visual Studio 2019` image with Docker/WSL/Hyper-V installed. You can try WSL on `Visual Studio 2019` image: https://www.appveyor.com/docs/windows-images-software/#wsl. Let me know how that worked!

**Comment by TheLastProject** [comment]
I'm having trouble actually getting anything to run under wsl: https://ci.appveyor.com/project/TheLastProject/pext/builds/28051440#L178

```
wsl bash prepare_activate_translations.sh 70 https://hosted.weblate.org/exports/stats/pext/?format=json
Calls one batch program from another.
CALL [drive:][path]filename [batch-parameters]
  batch-parameters   Specifies any command-line information required by the
                     batch program.
If Command Extensions are enabled CALL changes as follows:
CALL command now accepts labels as the target of the CALL.  The syntax
is:
    CALL :label arguments
```

Am I doing something wrong? Isn't just running wsl supposed to work?

**Comment by TheLastProject** [comment]
Ah, prefixing it wish `ps: ` works. So, if I use `ps: wsl bash prepare_activate_translations.sh 70 https://hosted.weblate.org/exports/stats/pext/?format=json` in my .appveyor.yml all is good.

**Comment by FeodorFitsner** [comment]
Oh, good to know, thanks for the update!

**Comment by khs1994** [comment]
Can update docker CE to 19.03 on `Windows Server 2019` ? current is 18.09

**Comment by FeodorFitsner** [comment]
@khs1994 I've added a separate issue for that: https://github.com/appveyor/ci/issues/3229 We are trying to consolidate both `Windows Server 2019` and `Visual Studio 2019` images, so it might be better to have Docker CE on VS 2019 image instead Docker EE.

**Comment by MeFisto94** [comment]
This can be closed now that WSL is part of the Visual Studio 2019 environment, right?

**Comment by FeodorFitsner** [comment]
Exactly, thanks!
